### PR TITLE
[inductor] eliminate atomic_add when omp parallel is not enabled

### PR DESCRIPTION
**Pull request by xndcn**
By introducing macro OMP_IN_PARALLEL, atomic_add can be eliminated without omp parallel sections.

microbenchmarks/inductor_cpu_atomic.py shows slightly improved performance (shape of [1, 4096] do not trigger omp parallel):
```
before:
shape; type; torch scatter_add; inductor scatter_add; torch scatter_add (worst case); inductor scatter_add (worst case)
[1, 4096]; torch.float32; 0.005643149997922592; 0.02442546000565926; 0.015422269998452977; 0.02939044999948237
[1, 4096]; torch.int32; 0.005688549999831594; 0.02786582999760867; 0.0085422399934032; 0.02361671000471688
```
```
after:
[1, 4096]; torch.float32; 0.005414750003183144; 0.015056980000736075; 0.014075709996177466; 0.026262230003339937
[1, 4096]; torch.int32; 0.0053910099995846394; 0.01520503999927314; 0.0052869500086671906; 0.014212599999154918
```

(`torch scatter_add` performance improves greatly, ~~must because I have switched from gcc to clang~~ because I happened to build in DEBUG :joy: )

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

### CODE REVIEW

File: torch/_inductor/codegen/cpp.py [hunk 1]
```diff
@@ -1111,15 +1111,21 @@ def store(self, name, index, value, mode=None):
...
-            line = f"{var}[{cexpr_index(index)}] = {value};"
+            lines = [f"{var}[{cexpr_index(index)}] = {value};"]
         elif mode == "atomic_add":
             if not config.cpp.dynamic_threads and self.num_threads == 1:
-                line = f"{var}[{cexpr_index(index)}] += {value};"
+                lines = [f"{var}[{cexpr_index(index)}] += {value};"]
             else:
-                line = f"atomic_add(&{var}[{cexpr_index(index)}], {value});"
+                lines = [
+                            "#if defined(OMP_IN_PARALLEL) && OMP_IN_PARALLEL",
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:1120]: Can you remind me what the macro `OMP_IN_PARALLEL` mean? It is not defined in the standard, e.g., https://www.openmp.org/wp-content/uploads/OpenMP-API-Specification-5-2.pdf

Comment (xndcn) [reply torch/_inductor/codegen/cpp.py:1120]: No, it is not defined in the standard, so I add #define OMP_IN_PARALLEL manually in WorkSharing parallel function

File: benchmarks/dynamo/microbenchmarks/inductor_cpu_atomic.py [hunk 2]
```diff
@@ -71,8 +71,8 @@ def test_total_time(shapes, types):
...
 test total time
 shape; type; torch scatter_add; inductor scatter_add; torch scatter_add (worst case); inductor scatter_add (worst case)
-[1, 4096]; torch.float32; 0.14733232000025964; 0.05388864999986254; 0.1451428800010035; 0.06496850000075938
-[1, 4096]; torch.int32; 0.1440268700002889; 0.05882900999949925; 0.1429359899998417; 0.07036211000013282
-[1, 65536]; torch.float32; 1.3435545300012564; 0.15207924000151252; 1.2523296799986383; 3.1408327299982375
-[1, 65536]; torch.int32; 1.3407247500003905; 0.12999147000073208; 1.2956029100018895; 0.853825209999286
+[1, 4096]; torch.float32; 0.0052365999999892665; 0.017564740001034806; 0.013978460001453641; 0.024368539998249616
+[1, 4096]; torch.int32; 0.005814169999212027; 0.016935479998210212; 0.005414689994722721; 0.017164559994853335
+[1, 65536]; torch.float32; 0.1297304199943028; 0.13822156000060204; 0.24826899999425223; 3.4930834800070443
+[1, 65536]; torch.int32; 0.126586229998793; 0.1042632999997295; 0.10594573999696877; 1.0463533599977382
```
Comment (jgong5) [inline benchmarks/dynamo/microbenchmarks/inductor_cpu_atomic.py:77]: Do you mean the perf speedup of eager is from the clang compiler you were using? What compiler did you use for inductor codegen? Just curious why does the perf numbers differ so much. Inductor's perf is much slower than eager for small shapes. I guess we'd better understand the gap first. cc @zhuhaozhe

Comment (xndcn) [reply benchmarks/dynamo/microbenchmarks/inductor_cpu_atomic.py:77]: @jgong5 Sorry, after some investigation I found that I had the lower speed of eager due to the debug build in my last commit ( :joy: I have no idea when I have set the DEBUG env last time)
So the perf speedup is due to Release vs Debug build.

**Pull request closed without merging**